pkg/apis/volumesnapshot/v1beta1/types.go

```diff
@@ -0,0 +1,384 @@
+	// If not specified, the default snapshot class will be used if one exists.
+	// If not specified, and there is no default snapshot class, dynamic snapshot creation will fail.
+	// Empty string is not allowed for this field.
+	// TODO(xiangqian): a webhook validation on empty string.
+	// More info: https://kubernetes.io/docs/concepts/storage/volume-snapshot-classes
+	// +optional
+	VolumeSnapshotClassName *string `json:"volumeSnapshotClassName,omitempty" protobuf:"bytes,2,opt,name=volumeSnapshotClassName"`
+}
+
+// VolumeSnapshotSource specifies whether the underlying snapshot should be
+// dynamically taken upon creation or if a pre-existing VolumeSnapshotContent
+// object should be used.
+// Exactly one of its members must be set.
+// Members in VolumeSnapshotSource are immutable.
+// TODO(xiangqian): Add a webhook to ensure that VolumeSnapshotSource members
+// will not be updated once specified.
+type VolumeSnapshotSource struct {
+	// persistentVolumeClaimName specifies the name of the PersistentVolumeClaim
+	// object in the same namespace as the VolumeSnapshot object where the
+	// snapshot should be dynamically taken from.
+	// This field is immutable.
+	// +optional
+	PersistentVolumeClaimName *string `json:"persistentVolumeClaimName,omitempty" protobuf:"bytes,1,opt,name=persistentVolumeClaimName"`
+
+	// volumeSnapshotContentName specifies the name of a pre-existing VolumeSnapshotContent
+	// object.
+	// This field is immutable.
+	// +optional
+	VolumeSnapshotContentName *string `json:"volumeSnapshotContentName,omitempty" protobuf:"bytes,2,opt,name=volumeSnapshotContentName"`
+}
+
+// VolumeSnapshotStatus is the status of the VolumeSnapshot
+type VolumeSnapshotStatus struct {
+	// boundVolumeSnapshotContentName represents the name of the VolumeSnapshotContent
+	// object to which the VolumeSnapshot object is bound.
+	// If not specified, it indicates that the VolumeSnapshot object has not been
+	// successfully bound to a VolumeSnapshotContent object yet.
+	// NOTE: Specified boundVolumeSnapshotContentName alone does not mean binding
+	//       is valid. Controllers MUST always verify bidirectional binding between
+	//       VolumeSnapshot and VolumeSnapshotContent to avoid possible security issues.
+	// +optional
+	BoundVolumeSnapshotContentName *string `json:"boundVolumeSnapshotContentName,omitempty" protobuf:"bytes,1,opt,name=boundVolumeSnapshotContentName"`
+
+	// creationTime, if specified, represents the timestamp when the point-in-time
+	// snapshot was successfully cut on the underlying storage system.
+	// In dynamic snapshot creation case, it will be filled in upon snapshot creation
+	// with the "creation_time" field returned from CSI "CreateSnapshot" gRPC call.
+	// For a pre-existing snapshot, it will be set to the "size_byte" value returned
+	// from CSI "ListSnapshots" gRPC call if there exists a matching CSI driver that
```

Comment:
"from the CSI "ListSnapshots" gRPC call if the driver supports it."

Ditto everywhere

Reply:
updated.